geo/src/algorithm/linestring_segment.rs

```diff
@@ -0,0 +1,186 @@
+        let n_lines = self.lines().count();
+
+        // Return None if n is 0 or the maximum usize
+        if (n == usize::MIN) || (n == usize::MAX) {
+            return None;
+        } else if n > n_lines {
+            let total_len = self.euclidean_length();
+            let densified = self.densify(total_len / (n as f64));
+            return densified.line_segmentize(n);
+            // return Some(MultiLineString::new(vec![densified]))
+        } else if n_lines == n {
+            // if the number of line segments equals n then return the
+            // lines as LineStrings
+            let lns = self
+                .lines_iter()
+                .map(LineString::from)
+                .collect::<Vec<LineString>>();
+
+            return Some(MultiLineString::new(lns));
+        } else if n == 1 {
+            let mlns = MultiLineString::from(self.clone());
+            return Some(mlns);
+        }
+
+        // Convert X into an iterator of `Lines`
+        let lns = self.lines_iter().peekable();
+
+        // Vec to allocate the  new LineString segments Coord Vec
+        // will be iterated over at end to create new vecs
+        let mut res_coords: Vec<Vec<Coord>> = Vec::with_capacity(n);
+
+        // calculate total length to track cumulative against
+        let total_length = self.euclidean_length().abs();
+
+        // tracks total length
+        let mut cum_length = 0_f64;
+
+        // calculate the target fraction for the first iteration
+        // fraction will change based on each iteration
+        let segment_prop = (1_f64) / (n as f64);
+        let segment_length = total_length * segment_prop;
+
+        // instantiate the first Vec<Coord>
+        let mut ln_vec: Vec<Coord> = Vec::new();
+
+        // iterate through each line segment in the LineString
+        for (i, segment) in lns.enumerate() {
+            // first line string push the first coord immediately
+            if i == 0 {
+                ln_vec.push(segment.start)
+            }
+
+            let length = segment.euclidean_length().abs();
+
+            // update cumulative length
+            cum_length += length;
+
+            if (cum_length >= segment_length) && (i != (n_lines - 1)) {
+                let remainder = cum_length - segment_length;
+                // if we get None, we exit the function and return None
+                let endpoint = segment.line_interpolate_point((length - remainder) / length)?;
+
+                // add final coord to ln_vec
+                ln_vec.push(endpoint.into());
+
+                // now we drain all elements from the vector into an iterator
+                // this will be collected into a vector to be pushed into the
+                // results coord vec of vec
+                let to_push = ln_vec.drain(..);
+
+                // now collect & push this vector into the results vector
+                res_coords.push(to_push.collect::<Vec<Coord>>());
+
```

Comment:
Is the `drain` followed by `collect` intended to reuse `ln_vec`s allocation?

Reply:
Correct. I had to (or, didn't know another way to) allocate the vector outside of the forloop so elements in the vector could persist across iterations. The checks inside the loop determined when to push the Coords.